extmod/modbtree: Use buffer protocol for keys/values.

dlech opened the pull request:
This changes the btree implementation to use the buffer protocol for reading key/values in all methods. As pointed out in #8748, `str` and `bytes` objects are not the only bytes-like objects that could be used.

dlech committed:
extmod/modbtree: Use buffer protocol for keys/values.

This changes the btree implementation to use the buffer protocol for
reading key/values in all methods. As pointed out in [1], `str` and
`bytes` objects are not the only bytes-like objects that could be used.

[1]: https://github.com/micropython/micropython/issues/8748

Signed-off-by: David Lechner <david@pybricks.com>

AmirHmZz commented:
@dlech I think it would be better to modify docs too.

dpgeorge commented:
This looks good, a nice generalisation.

Can you please add a test, eg in `tests/extmod/btree1.py`?

dlech committed:
tests/extmod/btree1: Add non-bytes keys and values.

The btree implementation has been genereralized so that it takes any
object that implements the buffer protocol. This modifies the existing
tests to use strings and memoryviews to exercise this functionality.

Signed-off-by: David Lechner <david@pybricks.com>

dlech commented:
> Can you please add a test, eg in `tests/extmod/btree1.py`?

done.

dpgeorge commented:
Thanks for adding the test.  Merged in a565811f23c675ccbeed636c065828cdaf5fa96a
